src/sims/nic/e1000_bm/gem5/ip.h

```diff
@@ -0,0 +1,457 @@
+#define IP_OPT_TS	 (4|IP_OPT_DEBMEAS)	/* timestamp */
+#define IP_OPT_ESEC	 (5|IP_OPT_COPY)	/* DoD extended security */
+#define IP_OPT_CIPSO	 (6|IP_OPT_COPY)	/* commercial security */
+#define IP_OPT_RR	  7			/* record route */
+#define IP_OPT_SATID	 (8|IP_OPT_COPY)	/* stream ID (obsolete) */
+#define IP_OPT_SSRR	 (9|IP_OPT_COPY)	/* strict source route */
+#define IP_OPT_ZSU	 10			/* experimental measurement */
+#define IP_OPT_MTUP	 11			/* MTU probe */
+#define IP_OPT_MTUR	 12			/* MTU reply */
+#define IP_OPT_FINN	(13|IP_OPT_COPY|IP_OPT_DEBMEAS)	/* exp flow control */
+#define IP_OPT_VISA	(14|IP_OPT_COPY)	/* exp access control */
+#define IP_OPT_ENCODE	 15			/* ??? */
+#define IP_OPT_IMITD	(16|IP_OPT_COPY)	/* IMI traffic descriptor */
+#define IP_OPT_EIP	(17|IP_OPT_COPY)	/* extended IP, RFC 1385 */
+#define IP_OPT_TR	(18|IP_OPT_DEBMEAS)	/* traceroute */
+#define IP_OPT_ADDEXT	(19|IP_OPT_COPY)	/* IPv7 ext addr, RFC 1475 */
+#define IP_OPT_RTRALT	(20|IP_OPT_COPY)	/* router alert, RFC 2113 */
+#define IP_OPT_SDB	(21|IP_OPT_COPY)	/* directed bcast, RFC 1770 */
+#define IP_OPT_NSAPA	(22|IP_OPT_COPY)	/* NSAP addresses */
+#define IP_OPT_DPS	(23|IP_OPT_COPY)	/* dynamic packet state */
+#define IP_OPT_UMP	(24|IP_OPT_COPY)	/* upstream multicast */
+#define IP_OPT_MAX	 25
+
+#define IP_OPT_COPIED(o)	((o) & 0x80)
+#define IP_OPT_CLASS(o)		((o) & 0x60)
+#define IP_OPT_NUMBER(o)	((o) & 0x1f)
+#define IP_OPT_TYPEONLY(o)	((o) == IP_OPT_EOL || (o) == IP_OPT_NOP)
+
+/*
+ * Security option data - RFC 791, 3.1
+ */
+struct ip_opt_data_sec {
+        uint16_t	s;		/* security */
+        uint16_t	c;		/* compartments */
+        uint16_t	h;		/* handling restrictions */
+        uint8_t		tcc[3];		/* transmission control code */
+} __attribute__((__packed__));
+
+#define IP_OPT_SEC_UNCLASS	0x0000	/* unclassified */
+#define IP_OPT_SEC_CONFID	0xf135	/* confidential */
+#define IP_OPT_SEC_EFTO		0x789a	/* EFTO */
+#define IP_OPT_SEC_MMMM		0xbc4d	/* MMMM */
+#define IP_OPT_SEC_PROG		0x5e26	/* PROG */
+#define IP_OPT_SEC_RESTR	0xaf13	/* restricted */
+#define IP_OPT_SEC_SECRET	0xd788	/* secret */
+#define IP_OPT_SEC_TOPSECRET	0x6bc5	/* top secret */
+
+/*
+ * {Loose Source, Record, Strict Source} Route option data - RFC 791, 3.1
+ */
+struct ip_opt_data_rr {
+        uint8_t		ptr;		/* from start of option, >= 4 */
+        /* uint32_t	iplist __flexarr; /1* list of IP addresses *1/ */
+        uint32_t	iplist; /* list of IP addresses */
+    // TODO(HE7086): flexarr
+} __attribute__((__packed__));
+
+/*
+ * Timestamp option data - RFC 791, 3.1
+ */
+struct ip_opt_data_ts {
+        uint8_t		ptr;		/* from start of option, >= 5 */
+#if DNET_BYTESEX == DNET_BIG_ENDIAN
+        uint8_t		oflw:4,		/* number of IPs skipped */
+                        flg:4;		/* address[ / timestamp] flag */
+#elif DNET_BYTESEX == DNET_LIL_ENDIAN
+        uint8_t		flg:4,
+                        oflw:4;
+#endif
+        /* uint32_t	ipts __flexarr;	/1* IP address [/ timestamp] pairs *1/ */
```

Comment:
__flexarr is a gnu/gcc extension which is not supported in inherited classes in cpp